Primer monorepo

broccolini opened the pull request:
## Moving Primer into a mono repo

This pr works towards moving the source files for Primer back into the Primer org, to be pulled into GitHub via npm. More [context on this decision here](https://github.com/github/design-systems/issues/275#issuecomment-290666882), and here's an interesting post on [why Babel is a monorepo](https://github.com/babel/babel/blob/master/doc/design/monorepo.md).

- All packages will live in one repo to make it easier to make updates across Primer
- Individual packages and meta-packages will still be independently versioned and distributed via npm

**To-do:**
- [x] Remove git logs from packages
- [x] Add ignore for npm-debug.log (and remove any existing logs)
- [x] Remove individual module contributing guidelines and add one to the root directory
- [x] Update git and issue urls to point to `primer-css`
- [x] Setup [Lerna](https://github.com/lerna/lerna) for easier multi-package management
- [x] Check tests still work (esp primer-module-build)
- [x] Update Travis config
- [x] Create pre-release for testing with new style guide docs and GitHub
- [ ] Switch GitHub over to pulling in primer, remove and re-org stylesheets directory https://github.com/github/github/pull/74495

**Post-ship to-do:**
- Add installation instructions for design-systems so that they can run Lerna etc. and publish releases
- Figure out how to manage global and individual package and meta-package changelogs (perhaps using [lerna/changelog](https://github.com/lerna/lerna-changelog))
- Setup github releases (perhaps using [zeit/release](https://github.com/zeit/release)) 
- Deprecate old packages and their repos
- Add doc explaining decision for mono repo (like Babel does)
- Explore using [Semantic Release](https://github.com/semantic-release/semantic-release) to help us better manage versions and the changelog (I think this is low priority but would like to see what @jonrohan and Shawn think once we've used the new setup with Lerna for a while)

Questions:
- ~~Should each package and meta-package have it's own `.gitignore` file or just the root?~~ - yes
- ~~Should each package and meta-package have it's own LICENSE or just the root?~~ - yes

@primer/design-systems @muan @josh @mislav What other things do I need to take into account? Any questions, concerns? Let me know if you have any thoughts on using Lerna and Semantic Release—they look like good options and help us with some of our maintenance and publishing workflow issues. I'll need help once this is ready with switching GitHub over to pulling this in a dependency.

muan commented:
So excited for the new setup! :sparkles:

> * Should each package and meta-package have it's own `.gitignore` file or just the root?
> * Should each package and meta-package have it's own LICENSE or just the root?

I think they should have their own ones since each package can be installed separately. User will need both files when they are using just one package.

> [ ] Setup [Lerna](https://github.com/lerna/lerna) for easier multi-package management

:+1: on this

> [ ] Explore using [Semantic Release](https://github.com/semantic-release/semantic-release) to help us better manage versions and the changelog

IMO this doesn't give us huge wins since `lerna` already prompts us to enter a version number so this is already a deliberate + essential action.

> [ ] Create pre-release for testing with new style guide docs and GitHub

With normal one-to-one repo/npm packages, you would be able to just install from source without a release. Like `npm install github:muan/romanize-names#test-branch`, this was the way I was suggesting to do testing with. I wasn't able to find an example like this for `lerna`. Have you come across one? is pre-release the only way?

Also it looks to me that packages _aren't_ versioned individually for React, Babel, and Jest, and I can't find `package.json` for each package in Angular, Meteor, or Ember. Do you know of an example repo somewhere?

broccolini commented:
Thanks @muan for the feedback! Replies below but also going to follow up with a comment re breaking this up a bit since not everything is essential for the first ship to make this a mono-repo.

> I think they should have their own ones since each package can be installed separately. User will need both files when they are using just one package.

🆒 

> IMO this doesn't give us huge wins since lerna already prompts us to enter a version number so this is already a deliberate + essential action.

The value of semantic release seems to be removing the human element from making decisions about version changes. Folks incorrectly bumping the version number has been a continuous issue, I think due to a combination of personal opinion, being unfamiliar with Semver, or because mapping CSS changes to Semver isn't always obvious. I'm going to leave this til later though as I agree that it's not essential to shipping this pr, and Lerna will stop us bumping a minor or major number and forgetting to reset the patch to 0. 😬 

> With normal one-to-one repo/npm packages, you would be able to just install from source without a release. Like npm install github:muan/romanize-names#test-branch, this was the way I was suggesting to do testing with. I wasn't able to find an example like this for lerna. Have you come across one? is pre-release the only way?

I'd like to chat with you about this. I suggested this since I thought was a normal/best practice since it keeps the distribution via npm and allows you to share with the open-source community for testing. Maybe this isn't necessary for us, but want to make sure it's easy for all GitHub teams using Primer. This will increase over the next year as more non-dotcom sites move over to Primer so whatever we do we need to make sure it's easy to communicate and test updates.

> Also it looks to me that packages aren't versioned individually for React, Babel, and Jest, and I can't find package.json for each package in Angular, Meteor, or Ember. Do you know of an example repo somewhere?

Not without spending some time searching around, no. But it doesn't matter to me that those projects aren't individually versioning their packages, it's a supported feature of Lerna and seems like the right thing to do for us, and it maintains our current versioning system. It makes it easier to push individual updates more safely since folks using it will be able to specify what they're ready to update to. I.e. if we updated our button styles and our box styles, but box styles would cause a lot of work for a team to update to, they'd still be able to pull in the new button styles and only update their box styles when they had the time to. If we weren't using Lerna I'd still want to set the repo up this way, Lerna just makes it easier.

Also re working on this in dev (in case that's what you were curious about), Lerna handles symlinking and when we run the bootstrap script it will look at the changes logged in git and ask us what we want to bump the version number too automatically. So having individually versioned should be fine. I'm testing this in a another repo before blowing everything up here so will see how it goes!

muan commented:
> Not without spending some time searching around, no. But it doesn't matter to me that those projects aren't individually versioning their packages, it's a supported feature of Lerna and seems like the right thing to do for us, and it maintains our current versioning system.

Sorry, I'm not doubting that we should do this, but asking what does this set up looks like in real life.

broccolini commented:
Updates and notes from conversations with @muan and @jonrohan:

- We don't think we'll be able to point to a branch in `github/github` when testing changes to Primer because of there is not git url mapped to the packages on npm, but we can publish a tagged pre-release. This will be fine once we have more components updated and in Primer but it might feel a little painful at first, and I think there's other things we can do to improve this new workflow (such as automating component testing or being able to spit out a list of all component variations).
- We're unsure why Greenkeeper stopped working, may be to do with the authentication changes a while back, will ask @jonrohan for help with setting that up again.
- Did we use ava for testing previously? If so, do we need to do anything re setup with the new monorepo? cc @jonrohan 
- I've updated the pr to reflect pre-ship and post-ship tasks so it's clear what we need to do for shipping this.

lenovouser commented on packages/README.md:
```diff
@@ -1,3 +1,3 @@
 # Primer modules
 
-Primer-css includes 23 modules which can be installed altogether or individually via npm. We also group our modules into 3 meta-packages which you can install individually: Primer-core, Primer-product, and Primer-marketing.
+Primer-css is monorepo composed of 23 packages which can be installed altogether or individually via NPM. We group our packages into 3 meta-packages which you can install individually: [primer-core](primer-core), [primer-product](primer-product), and [primer-marketing](primer-marketing). Install [primer-css](primer-css) if you want to install all 23 packages.
```
```diff
+Primer-CSS
-Primer-css
```

might be better 🙂

broccolini commented:
Notes from meeting with @muan and @jonrohan 

> We're unsure why Greenkeeper stopped working, may be to do with the authentication changes a while back, will ask @jonrohan for help with setting that up again.

- Probably stopped working after updating permissions. Don't really need it anymore since Lerna handles keeping the packages up to date. Could use it just for keeping Lerna up to date if we want to.

> Did we use ava for testing previously? If so, do we need to do anything re setup with the new monorepo? cc @jonrohan

- Don't really need ava for CSS, more useful for JS.

broccolini commented:
@josh @mislav, and @dgraham 👋 I've been synching up with @muan about this (and @jonrohan now that he's back!), but wanted to make sure you had time to review this since we're getting close to shipping.

Some things to note (and shorter summary of initial pr comment):
- Primer will be pulled into github/github as a dependency, I'll open a pr for that shortly for testing and will request your review
- Primer will now be a monorepo, and [Lerna](https://github.com/lerna/lerna) will be a dependency for this repo because we are using it for publishing and versioning multiple packages
- We now have just one travis config in the root directory which runs through and tests all the packages. See the [travis config here](https://github.com/primer/primer-css/pull/230/files#diff-354f30a63fb0907d4ad57269548329e3), and [here's the output](https://travis-ci.org/primer/primer-css). If there is a better way to approach that let me know.

---
@primer/design-systems et al, if you're interested in testing the updates in this branch, here's the versions you'll need to install:
- `primer-css@7.0.0-0` (all of the packages)
- `primer-core@4.0.0-0` (core primer packages only, currently @imported into `frameworks.scss` and `site.scss`)
- `primer-product@4.0.0-0` (product packages only, currently @imported into `github.scss`)
- `primer-marketing@4.0.0-0` (marketing packages only, currently @imported `site.scss`)

This should match everything that's in github/github right now, including module updates that haven't yet been publish to the Primer repo. 

Here's is a [gist showing the unpublished changes](https://gist.githubusercontent.com/broccolini/6bb467806cc3f2a85f07ffeb64a3e0c6/raw/b7b787ad766fccdc2d8ca909422815b9461afe86/unpublished-primer-changes.md) for comparison - @jonrohan it would be rad if you could double check and review those changes. I've bumped to pre-patch/minor/major which should match except for `primer-layout` where there have been 2 major bumps that haven't been published but both are included in this pr so I've just bumped it once.

broccolini commented:
**Updates:**
Discussed with @jonrohan and we've decided to remove core as a peer dependency from marketing and product meta-packages. It's not really a peer dependency, each package works entirely without each other. In some documentation we use components from core in other components, but they are not actually required to use the individual components. It also often causes more problems since npm changed how peer dependencies are installed so you get a bunch of errors and then have to decide whether to install those packages. Basically seems like it's not worth it.

I also realized I missed `primer-marketing-support` from the initial monorepo setup, so have added that now. Here are the new package versions for testing:

- `primer-css@7.0.0-1`
- `primer-marketing@4.0.0-1`
- `primer-product@4.0.0-1`
- `primer-core@4.0.0-0` is unchanged from the last release

@sophshep As discussed, would be rad if you can test on some marketing sites.

WIP dotcom pr for switching Primer to a dependency here: https://github.com/github/github/pull/74495

broccolini commented:
Thanks @mislav! Currently the builds take ~4 minutes. I feel that that's not too long however we can look into caching as you suggested if it ends up becoming a problem.

broccolini commented:
@sophshep Re testing, I'd like your help checking that you can point your marketing sites to a pre-release version. After thinking this through though, I reckon it would be good to jump on a call and see what your workflow is with Jekyll/non dotcom sites and then I can help explain what, if any, changes to that workflow would be with this new setup. If you can make time to do this tomorrow that would be rad - feel free to ping me in slack re times to sync.

broccolini commented:
**Update:** I published a new pre-release to include dotcom updates to [grid gutters](https://github.com/github/github/pull/74164/files#diff-9570733c684e55259f392d82a8644050), and [fade variables](https://github.com/github/github/pull/74179). I've asked for a freeze on publishing updates until this is shipped, unless there's a serious bug that needs fixing. Hoping to get this shipped Wednesday if everyone has enough time to test.

Here are the new package versions for testing:

- `primer-css@7.0.0-2`
- `primer-marketing@4.0.0-2`
- `primer-product@4.0.0-2`
- `primer-core@4.0.0-1`

jonrohan commented on packages/primer-forms/lib/form-validation.scss:
```diff
@@ -0,0 +1,398 @@
+    display: none;
+  }
+}
+
```
this is missing this line https://github.com/github/github/blob/master/app/assets/stylesheets/primer-core/forms/lib/form-validation.scss#L119-L122

```
.repo-config-option:not(:checked) ~ .time-remaining {
  display: none;
}
```

jonrohan commented on packages/primer-marketing/index.scss:
```diff
@@ -0,0 +1,23 @@
+@import "primer-cards/index.scss";
+@import "primer-page-headers/index.scss";
+@import "primer-page-sections/index.scss";
+@import "primer-tables/index.scss";
```
Missing an import for `"primer-marketing-utilities/index.scss"`

broccolini commented:
@shawnbot Here's the error I'm getting:

<img width="1255" alt="screenshot 2017-06-22 18 03 43" src="https://user-images.githubusercontent.com/334891/27489954-70fcc428-580a-11e7-8d6b-e356a36f0a71.png">

This was after I updated the marketing package to include the import for `primer-marketing-utilities` and also added that as a dependency. After updating packages, I ran `lerna bootstrap` to symlink the new dependency. Then ran `lerna publish` and got the above errors.

I thought maybe it was just having an issue with previous versions of how the module were setup, so I ran `lerna clean` to clean out the node modules, and then re-ran `lerna bootstrap` to symlink all the packages again. This time `lockfiles` were added to each package which I removed as previously this seemed to cause an issue with Lerna, but I kept everything else the same. After running `lerna publish` again now I get the same error above on other packages:
<img width="1276" alt="screenshot 2017-06-23 12 03 09" src="https://user-images.githubusercontent.com/334891/27490358-f7c002e4-580b-11e7-9eba-dd0bf13f4771.png">

I manually removed the node modules and tried this all over again, to no avail. I've tried manually installing packages and I tried this with and without lockfiles. I started to look through [issues in the Lerna repo](https://github.com/lerna/lerna/issues), and found that [lerna isn't compatible with npm @5](https://github.com/lerna/lerna/issues/856#issuecomment-304959008), which I am on. So that may be why I'm getting a bunch of issues 😑 . 

cc @jonrohan

hzoo commented:
I guess this is rather late but I can answer questions y'all might have about Lerna (I maintain Babel and Lerna somewhat). Not sure how well it works with npm 5 and hoisting since never looked into it, but does seem like everyone did their research pretty well 😄 (lerna slack: https://slack.lernajs.io/, babel: slack.babeljs.io)

broccolini commented:
Thanks @hzoo! I had joined the lerna slack but @shawnbot ended up helping me solve the issues we were having. There's some follow up work to do on our side and there's some feedback we'll pass on to the you and the Lerna maintainers if issues haven't been opened already.

I don't think the problems we hit were really npm@5 issues, however it didn't seem to like the lockfile. Either way, since there are so many bug reports for npm@5 at the moment, it seems sensible to stick with 4 for now.

Also, you're more than welcome to randomly comment on our Primer pr's 😃
